<a href="https://colab.research.google.com/github/raahatg21/Movie-Recommender-System/blob/master/simple_recommender.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Simple Recommender System of Movies

Simple recommenders offer generalised recommendations to every user, based on movie popularity and/or genre. The basic idea behind this system is that movies that are more popular and critically acclaimed will have a higher probability of being liked by the average audience. 

The output will be same for all users and all products. We use the dataset of all movies in IMDB (derived by MovieLens 20M Dataset).

### Workflow
* Load the data of all movies on IMDB
* Select a threshold as minimum number of votes, and reject all movies below that votes
* Select a metric for ranking and sort the list based on that metric
* Output top 15 movies of the sorted list

In [0]:
import pandas as pd
import os

# Yes, that's it! Ain't much going on here.

In [0]:
# Mounting Google Drive

from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=email%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdocs.test%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.photos.readonly%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/drive


In [0]:
BASE_DIR = '/content/drive/My Drive/Recommender System/the-movies-dataset'

In [0]:
metadata_path = os.path.join(BASE_DIR, 'movies_metadata.csv')
credits_path = os.path.join(BASE_DIR, 'credits.csv')
keywords_path = os.path.join(BASE_DIR, 'keywords.csv')

In [0]:
# Load the movies metadata

metadata = pd.read_csv(metadata_path, low_memory = False)

metadata.shape

(45466, 24)

In [0]:
metadata.head(5)

adult  ... vote_count
0  False  ...     5415.0
1  False  ...     2413.0
2  False  ...       92.0
3  False  ...       34.0
4  False  ...      173.0

[5 rows x 24 columns]

Our weighted rating metric is,
### WR = ( ( v / (v + m)) * R)  +  ( ( m / (v + m)) * C)

where,
- v is the number of votes for the movie
- m is the minimum votes required to be listed in the chart
- R is the average rating of the movie
- C is the mean vote across the whole report

In [0]:
C = metadata['vote_average'].mean()

print(C)

5.618207215133889


In [0]:
m = metadata['vote_count'].quantile(0.90)
# Choosing only the top 10% popular movies (as most movies in IMDB dataset are crap anyways)

print(m)

160.0


Thus, a movie can be in the top 10 even if it has greater than 160 votes.

In [0]:
# Filter all qualified movies and make another dataset out of them

q_movies = metadata.copy().loc[metadata['vote_count'] >= m]

q_movies.shape

(4555, 24)

Thus, only 10% of all movies remain

In [0]:
# Function that computes the weightted rating for each entry

def weighted_rating(x, m = m, C = C):
  v = x['vote_count']
  R = x['vote_average']
  
  score = (v / (v+m) * R) + (m / (m+v) * C)
  
  return score

In [0]:
q_movies['score'] = q_movies.apply(weighted_rating, axis = 1)  # Applying the function to every row

In [0]:
q_movies = q_movies.sort_values('score', ascending = False)

In [0]:
# Printing the details of top 15 movies

q_movies[['title', 'vote_count', 'vote_average', 'score']].head(15)

title  vote_count  vote_average     score
314           The Shawshank Redemption      8358.0           8.5  8.445869
834                      The Godfather      6024.0           8.5  8.425439
10309      Dilwale Dulhania Le Jayenge       661.0           9.1  8.421453
12481                  The Dark Knight     12269.0           8.3  8.265477
2843                        Fight Club      9678.0           8.3  8.256385
292                       Pulp Fiction      8670.0           8.3  8.251406
522                   Schindler's List      4436.0           8.3  8.206639
23673                         Whiplash      4376.0           8.3  8.205404
5481                     Spirited Away      3968.0           8.3  8.196055
2211                 Life Is Beautiful      3643.0           8.3  8.187171
1178            The Godfather: Part II      3418.0           8.3  8.180076
1152   One Flew Over the Cuckoo's Nest      3001.0           8.3  8.164256
351                       Forrest Gump      8147.0           8.2  8.150272
1154           The Empire Strikes Back      5998.0           8.2  8.132919
1176                            Psycho      2405.0           8.3  8.132715